### PR TITLE
feat: custom Registry port & fallbacks, spawn var fix

### PR DESCRIPTION
**Pull request by dtbuchholz**
## Summary

Tests will occasionally fail without insight, causing a mocha "exceeded timeout" error. With this, a `fallback` and `registryPort` option can now be set, allowing you to enable fallbacks and/or use a custom Registry port for hardhat. Also, the Registry spawn process wasn't using the passed env variables, which could be causing some errors.

Closes [LOT-6](https://linear.app/tableland/issue/LOT-6/ci-improvement-with-tests).

## Details

Port 8545 being already in use—which is what the Registry / hardhat node bind to—seems causes CI errors since hardhat is trying to connect to a taken port. This adds functionality for retry logic that probes port 8545 before spinning up the Registry.

- Created utils to probe / check port availability.
- Allow a `fallback` option (e.g., `npx local-tableland --fallback`) such that if the default port used is in use, it will attempt backup ports and spin up the Registry from there. If so, it'll log something like `[Notice] Registry default port in use, using fallback port 8546`.
- Allow a `registryPort` option (e.g.,  `npx local-tableland --fallback --registryPort 1234`) to set a custom port for hardhat. It'll log something like `[Notice] Registry is using custom port 1234`.
- Altered `getAccounts()` helper to, optionally, take an instance of LT to grab the port being used, in case a fallback occurs. (Will default to using port 8545 if not LT instance is passed, ensuring backwards compatability).
- Created a `getRegistryPort()` helper to get the Registry port number being used in case of fallbacks, also optionally taking LT as a param.
- Stored Registry spawn process variables to `process.env.HARDHAT_...` vs. defining/passing them directly to fix undefined vars being passed.
- Also added some additional error handling to, hopefully, improve insight into what happens when tests fail unexpectedly in mocha before/after hooks.

When you pass a custom port, or if a fallback occurs, the `validator/config.json` is temporarily overwritten. This allows the validator to use whatever port is set. With shutdown processes, this file gets reset to the original, ensuring port changes that happen during execution always return to the original state (i.e., port 8545) and don't cause side effects.

One callout is that you can't run two LT instances at the same time because they're sharing resources. The second instance will error out with something like `initializing db connection: creating migration: disk I/O error: no such file or directory`. But, this base functionality should enable future features like spinning up multiple hardhat nodes on different ports with a single validator, mimicking a multichain testing setup.

## How it was tested

- Added tests for general start and shutdown processes.
- Added tests to check how LT startup happens in the port in use scenario. E.g., if the port is taken for the _all_ of the retry attempts + fails, or if the port is taken temporarily and successfully triggers retries + actually deploys the Registry.
- Added tests for using a custom `registryPort`.
- Also created a rudimentary way to log the `start()` and `shutdown()` process execution times since they showed a high variance, particularly, for shutdown. Sometimes, its average/median was ~10 ms, but often, it's 10 seconds.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -73,6 +74,21 @@ class LocalTableland {
     // TODO: I don't think this is doing anything anymore...
     this.#_cleanup();
 
+    // check if the port is in use and retry 5 times (1 second between each)
```
Comment (dtbuchholz) [inline src/main.ts:77]: i figured no need for something like exponential backoff...simple probing every 1 second. can adjust if needed.

File: src/util.ts [hunk 2]
```diff
@@ -319,3 +320,22 @@ export const getAccounts = function (): Wallet[] {
     return wallet.connect(getDefaultProvider("http://127.0.0.1:8545"));
   });
 };
+
+export function checkPortInUse(port: number): Promise<boolean> {
```
Comment (dtbuchholz) [inline src/util.ts:324]: return true/false if the port is in use via checking for `EADDRINUSE`

File: test/cli.test.mjs [hunk 3]
```diff
@@ -1,61 +1,77 @@
 import chai from "chai";
 import shell from "shelljs";
+import { join } from "path";
 
 const expect = chai.expect;
 
 describe("Project builder", function () {
+  // Store absolute paths to the current working directory & path to tmp dir
```
Comment (dtbuchholz) [inline test/cli.test.mjs:8]: these changes aren't required, but when doing local-only GH `test.yml` checks with `act`, i ran into issues with its usage of docker + relative paths. i just make the paths absolute.

File: test/e2e.test.mjs [hunk 4]
```diff
@@ -6,10 +6,81 @@ import {
...
+import { startMockServer, stopMockServer } from "./util.mjs";
 
 const expect = chai.expect;
 const localTablelandChainId = 31337;
 
+describe("Validator and Chain startup and shutdown", function () {
+  let server;
+  const lt = new LocalTableland({ silent: true });
+
+  this.timeout(20000); // Starting up LT takes 3000-7000ms; shutting down takes <10-10000ms
```
Comment (dtbuchholz) [inline test/e2e.test.mjs:18]: i observed that on my machine, the LT startup usually takes 3-7 seconds, and shutdown was <10ms to 10 seconds. i'm not sure why the variance. 

e.g., one test would log:
```
  Validator and Chain startup and shutdown
lt start took 4313 milliseconds.
    ✔ fails to start due to being blocked on port 8545 (11328ms)
stopMockServer took 0 milliseconds.
lt.shutdown took 8 milliseconds.
lt start took 6968 milliseconds.
    ✔ successfully starts using retry logic after initially blocked (6968ms)
stopMockServer took 0 milliseconds.
lt.shutdown took 7 milliseconds.
```
and running it again would log:
```
  Validator and Chain startup and shutdown
lt start took 3371 milliseconds.
    ✔ fails to start due to being blocked on port 8545 (10384ms)
stopMockServer took 0 milliseconds.
lt.shutdown took 9 milliseconds.
lt start took 5846 milliseconds.
    ✔ successfully starts using retry logic after initially blocked (5847ms)
stopMockServer took 0 milliseconds.
lt.shutdown took 10021 milliseconds.
```

File: test/e2e.test.mjs [hunk 5]
```diff
@@ -19,12 +90,20 @@ describe("Validator, Chain, and SDK work end to end", function () {
   // These tests take a bit longer than normal since we are running them against an actual network
   this.timeout(20000);
   before(async function () {
-    await lt.start();
-    await new Promise((resolve) => setTimeout(() => resolve(), 2000));
+    try {
+      await lt.start();
+      await new Promise((resolve) => setTimeout(() => resolve(), 2000));
+    } catch (err) {
+      expect.fail(`failed to start local network: ${err.message}`);
```
Comment (dtbuchholz) [inline test/e2e.test.mjs:97]: i figured adding this additional handling can't hurt

File: src/main.ts [hunk 6]
```diff
@@ -73,18 +74,38 @@ class LocalTableland {
...
+    }
+    // throw if all retries failed and port is in use
+    if (retries === 5) {
+      throw new Error(
+        `cannot start a local chain due to port ${hardhatPort} in use`
+      );
+    }
+
+    // You *must* store these in `process.env` to access within the hardhat subprocess
+    process.env.HARDHAT_NETWORK = "hardhat";
```
Comment (dtbuchholz) [inline src/main.ts:95]: here, i directly store the variables in process.env. if you don't do this and only pass them in the spawned child process, for some reason, they are never read (i.e., undefined within the hardhat scripts).

File: src/main.ts [hunk 7]
```diff
@@ -73,18 +74,38 @@ class LocalTableland {
...
+    process.env.HARDHAT_UNLIMITED_CONTRACT_SIZE = "true";
+    process.env.HARDHAT_PORT = hardhatPort.toString();
+
     // Run a local hardhat node
     this.registry = spawn(
       isWindows() ? "npx.cmd" : "npx",
-      ["hardhat", "node"],
+      ["hardhat", "node", "--port", hardhatPort.toString()], // TODO: use a fallback port on conflicts vs. failing on retries
```
Comment (dtbuchholz) [inline src/main.ts:102]: this doesn't change any behavior but at least provides the option to use custom port fallbacks, if desired in the future.

File: test/e2e.test.mjs [hunk 8]
```diff
@@ -1,30 +1,317 @@
...
+        lt.shutdown()
+      );
+      executionTimes.shutdown.push(shutdownExecutionTime);
+    } catch (err) {
+      expect.fail(`failed to cleanup processes: ${err.message}`);
+    }
+
+    // Calculate & log the min, max, median, and average start and shutdown times
+    console.log(`\nExecution metrics`);
+    logMetrics("start()", executionTimes.start);
```
Comment (joewagner) [inline test/e2e.test.mjs:313]: This logging is great!

File: src/util.ts [hunk 9]
```diff
@@ -69,28 +92,88 @@ const configDescriptors: ConfigDescriptor[] = [
...
+   */
   silent?: boolean;
+
+  /**
+   * Use a fallback Registry port if the default port 8545 is in use or a custom
+   * port is desired. Note that clients will need to be configured to use this
+   * port over the default port, e.g., connect to `http://127.0.0.1:<fallbackPort>`
+   * instead of port 8545.
+   */
+  fallback?: boolean;
```
Comment (joewagner) [inline src/util.ts:157]: Does this work from a practical standpoint?  The chain will start on the fallback port, but how will the Validator know how to connect to the fallback?

For the sake of getting the CI tests to pass it seems easier to just wait until the standard port is available.  If we discover that can't work, we could create a non-standard validator config and retry starting with that and the `registryPort` options set.

File: src/util.ts [hunk 10]
```diff
@@ -69,28 +92,88 @@ const configDescriptors: ConfigDescriptor[] = [
...
+
+  /**
+   * Use a custom Registry hardhat port, e.g., `http://127.0.0.1:<registryPort>`.
+   * Must have fallbacks enabled.
+   */
+  registryPort?: number;
```
Comment (joewagner) [inline src/util.ts:163]: In general enabling a non-default registry port seems great!  Seems worth documenting that you'll have to use a non-default Validator config with the `validator` option. 
ref: https://github.com/tablelandnetwork/local-tableland/pull/423

File: .eslintrc.cjs [hunk 11]
```diff
@@ -14,13 +14,21 @@ module.exports = {
...
   extends: [
     "standard",
     "plugin:prettier/recommended",
     "plugin:node/recommended",
   ],
+  overrides: [
+    {
+      files: ["validator/config.json"],
+      rules: {
+        "eol-last": "off",
```
Comment (joewagner) [inline .eslintrc.cjs:28]: Why turn this off?

**Review by joewagner — CHANGES_REQUESTED** [review]
@dtbuchholz Nice work here!  
Lots of good changes, however I have a few concerns.  
I added inline comments, but If it's easier I'm happy to jump in a call. 
TL;DR; is two main things:
- the fallback port idea seems like it won't work in practice
- doesn't seem like using try-catch in tests is best practice.  This is certainly up for debate, but we don't do it anywhere else.  If there's a reason it needs to be added here, let's discuss further.

File: src/util.ts (hunk 10 above)
Comment (dtbuchholz) [reply src/util.ts:140]: note: created an issue for myself for adding docs on both the `validator` and `registryPort` options: [here](https://linear.app/tableland/issue/DOC-80/update-docs-with-custom-registry-port-and-validator-config)

**Comment by dtbuchholz** [comment]
@joewagner i've removed the fallback logic and should have your other recommendations incorporated (e.g., removed try/catch from certain test actions, removed unneeded prettier/linting EOL). lmk if i missed anything!

File: src/util.ts [hunk 12]
```diff
@@ -69,28 +89,68 @@ const configDescriptors: ConfigDescriptor[] = [
...
+  /**
+   * Instance of a Tableland Registry.
+   */
   registryDir?: string;
+  /**
+   * Path to the Tableland Registry contract repository.
+   */
   docker?: boolean;
+  /**
+   * Use Docker to run the Validator.
```
Comment (joewagner) [inline src/util.ts:127]: Should this comment be above the `docker` prop?

File: src/up.ts [hunk 13]
```diff
@@ -68,8 +77,12 @@ const go = async function () {
 
 // start a tableland network, then catch any uncaught errors and exit loudly
 go().catch((err) => {
-  console.error("unrecoverable error");
-  console.error(err);
+  if (err.message.includes(`port already in use`)) {
+    console.error(err.message);
```
Comment (joewagner) [inline src/up.ts:81]: Sorry for the nit here, but it would be nice to log the entire error object.  It makes debugging unexpected errors more difficult if you don't have a stack trace.

Comment (dtbuchholz) [reply src/up.ts]: i'm all for the nits! updated them.

**Review by joewagner — APPROVED** [review]
@dtbuchholz This is really awesome.  I tested it out with the JS monorepo and it's gonna let us run all the tests in parallel.  That alone is really nice, but if we can also fix the constant CI issues this is a huge win!
Since it's a bug fix I'd like to publish asap.  Sanity check here, does this change any default functionality? i.e. is this a patch version bump?

File: src/validators.ts [hunk 14]
```diff
@@ -76,22 +114,37 @@ class ValidatorPkg {
     for (const filepath of dbFiles) {
       shell.rm("-f", filepath);
     }
+
+    // reset the Validator config file in case it was modified with a custom
+    // Registry hardhat port
+    if (ORIGINAL_VALIDATOR_CONFIG) {
```
Comment (dtbuchholz) [inline src/validators.ts:120]: the original file ends with a new line, but the file gets overwritten with no new line char at the end. thus, running LT will show a file has changed, even though it really hasn't. two things come to mind:
- add a `\n` to the end of the file
- use an eol override for eslint and remove it from prettier (i dont _think_ it has a specific file/eol rule i can use)

**Comment by dtbuchholz** [comment]
@joewagner nice! and this adds the ability a new feature where you can pass a `--registryPort` flag, but everything else is the same.

one minor detail i just noticed something upon rewriting the validator config, which i _thought_ was a non-issue. see the comment above re: eslint/prettier and an ending return line.

File: src/validators.ts [hunk 15]
```diff
@@ -50,6 +58,25 @@ class ValidatorPkg {
       validatorUri = this.validatorDir;
     }
 
+    // get the validator config file
+    const configFilePath = join(this.validatorDir, "config.json");
+    const configFile = readFileSync(configFilePath);
+    const validatorConfig = JSON.parse(configFile.toString());
+
+    // save the validator config state
+    ORIGINAL_VALIDATOR_CONFIG = JSON.stringify(validatorConfig, null, 2);
```
Comment (joewagner) [inline src/validators.ts:67]: @dtbuchholz This is left over from when this package only worked with docker.  What if we change this line to:
`ORIGINAL_VALIDATOR_CONFIG = configFile.toString();`
Then we don't have to deal with changing the lint and prettier setting for all of our ts packages?

Comment (joewagner) [reply src/validators.ts:67]: If that makes sense, we should probably do the same thing on lone 159 too

Comment (dtbuchholz) [reply src/validators.ts]: @joewagner yes, nice solution! should be good to go now.

**Review by joewagner — APPROVED** [review]
Excellent!  really excited to push this out to all the tests 🚀
